%3a14/srfi-14.scm

```diff
@@ -1,804 +1,828 @@
+(define char-set-union
+  (make-char-set-op char-set-union! make-empty-char-set))
+
+(define char-set-intersection!
+  (make-char-set-op! simple-cset-intersection! inversion-list-intersection))
+(define char-set-intersection
+  (make-char-set-op char-set-intersection! make-full-char-set))
+
+(define char-set-difference!
+  (make-char-set-op! simple-cset-difference! inversion-list-difference))
+
+(define (char-set-difference cset1 . csets)
+  (apply char-set-difference! (char-set-copy cset1) csets))
+
+; copied from inversion-list.scm
+(define (binary->n-ary proc/2)
+  (lambda (arg-1 . args)
+    (if (and (pair? args)
+	     (null? (cdr args)))
+	(proc/2 arg-1 (car args))
+	(let loop ((args args)
+		   (result arg-1))
+	  (if (null? args)
+	      result
+	      (loop (cdr args) (proc/2 result (car args))))))))
+
+(define inversion-list-xor
+  (binary->n-ary
+   (lambda (i-list-1 i-list-2)
+     (inversion-list-union (inversion-list-intersection
+			    (inversion-list-complement i-list-1)
+			    i-list-2)
+			   (inversion-list-intersection
+			    i-list-1
+			    (inversion-list-complement i-list-2))))))
+
+; Really inefficient for things outside Latin-1
+; WHO NEEDS THIS NONSENSE, ANYWAY?
+(define char-set-xor!
+  (make-char-set-op! simple-cset-xor! inversion-list-xor))
+
+(define char-set-xor
+  (make-char-set-op char-set-xor! make-empty-char-set))
 
-(define (char-set-xor . csets)
-  (if (pair? csets)
-      (let ((s (%string-copy (%char-set:s/check (car csets) char-set-xor))))
-	(%char-set-algebra s (cdr csets) %xor! char-set-xor)
-	(make-char-set s))
-      (char-set-copy char-set:empty)))
+(define (char-set-diff+intersection! cs1 cs2 . csets)
+  (call-with-values
+      (lambda () (apply simple-cset-diff+intersection!
+			(char-set-simple cs1) (char-set-simple cs2)
+			(map char-set-simple csets)))
+    (lambda (simple-diff simple-intersection)
+      (set-char-set-simple! cs1 simple-diff)
+      (set-char-set-simple! cs2 simple-intersection)
+      (let ((i-list-1 (char-set-i-list cs1))
+	    (i-list-2 (char-set-i-list cs2))
+	    (i-list-rest (map char-set-i-list csets)))
+	(set-char-set-i-list! cs1
+			      (apply inversion-list-difference
+				     i-list-1 i-list-2
+				     i-list-rest))
+	(set-char-set-i-list! cs2
+			      (inversion-list-intersection
+			       i-list-1
+			       (apply inversion-list-union
+				      i-list-2
+				      i-list-rest)))
+	(values cs1 cs2)))))
 
+(define (char-set-diff+intersection cs1 . csets)
+  (apply char-set-diff+intersection!
+	 (char-set-copy cs1)
+	 (make-empty-char-set)
+	 csets))
 
-;;; -- Difference & intersection
+;; Byte vector utilities
 
-(define (%char-set-diff+intersection! diff int csets proc)
-  (for-each (lambda (cs)
-	      (%string-iter (lambda (i v)
-			      (if (not (zero? v))
-				  (cond ((si=1? diff i)
-					 (%set0! diff i)
-					 (%set1! int  i)))))
-			    (%char-set:s/check cs proc)))
-	    csets))
+(define (byte-vector-copy b)
```

Comment:
Wouldn't it be better to use a standard function for this instead of a custom one?

Reply:
If you think merging of new changes is a serious consideration, then I think it is fine to keep it, but if there aren't going to be a lot of changes, it might be worth it to clean up the code. However, I feel we can probably clean up the code "in tree" as it were, and it is probably fine to accept this at the moment.